crates/pallet-domains/src/lib.rs

```diff
@@ -157,6 +176,63 @@ mod pallet {
         },
     }
 
+    /// Per-domain state for tx range calculation.
+    #[derive(Debug, Default, Decode, Encode, TypeInfo, PartialEq, Eq)]
+    pub struct DomainState {
+        /// Current tx range.
+        pub tx_range: U256,
+
+        /// Blocks in the current adjustment interval.
+        pub interval_blocks: u64,
+
+        /// Bundles in the current adjustment interval.
+        pub interval_bundles: u64,
+    }
+
+    impl DomainState {
+        pub fn new() -> Self {
+            Self {
+                tx_range: Self::initial_tx_range(),
+                interval_blocks: 0,
+                interval_bundles: 0,
+            }
+        }
+
+        /// Called when a bundle is added to the current block.
+        pub fn on_bundle(&mut self) {
+            self.interval_bundles += 1;
+        }
+
+        /// Called when the current block is finalized.
+        pub fn on_finalize(&mut self) {
+            self.interval_blocks += 1;
+            if self.interval_blocks < TX_RANGE_ADJUSTMENT_INTERVAL_BLOCKS {
+                return;
+            }
+
+            // End of interval. Recalculate the tx range and reset the state.
+            let prev_tx_range = self.tx_range;
+            self.tx_range = calculate_tx_range(
+                self.tx_range,
+                self.interval_bundles,
+                EXPECTED_BUNDLES_PER_INTERVAL,
+            );
+            log::trace!(target: "runtime::domains",
+                "tx range update: blocks = {}, bundles = {}, prev = {prev_tx_range}, new = {}",
+                self.interval_blocks, self.interval_bundles, self.tx_range);
+            self.interval_blocks = 0;
+            self.interval_bundles = 0;
+        }
+
+        pub fn initial_tx_range() -> U256 {
+            U256::MAX / INITIAL_DOMAIN_TX_RANGE
+        }
+    }
+
+    #[pallet::storage]
+    pub(super) type DomainTxRangeState<T: Config> =
+        StorageMap<_, Identity, DomainId, DomainState, ValueQuery>;
```

Comment:
`OptionQuery` is more appropriate here, especially the default value does not make a lot of sense. And then in `update_tx_range_on_bundle` you don't have to check the existence and insert a new value before incrementing `interval_bundles`.

```suggestion
        StorageMap<_, Identity, DomainId, DomainState, OptionQuery>;
```

```diff
-    fn update_tx_range_on_bundle(domain_id: DomainId) {
-        if !DomainTxRangeState::<T>::contains_key(domain_id) {
-            DomainTxRangeState::<T>::insert(domain_id, DomainState::new());
-        }
-        DomainTxRangeState::<T>::mutate(domain_id, |domain_state| {
-            domain_state.on_bundle();
-        })
+    fn update_tx_range_on_bundle(domain_id: DomainId) {
+        DomainTxRangeState::<T>::mutate(domain_id, |maybe_domain_state| match maybe_domain_state {
+            Some(domain_state) => {
+                domain_state.interval_bundles += 1;
+            }
+            None => {
+                maybe_domain_state.replace(DomainState {
+                    tx_range: DomainState::initial_tx_range(),
+                    interval_blocks: 0,
+                    interval_bundles: 1,
+                });
+            }
+        });
     }
```